crates/bevy_reflect/src/path/mod.rs

```diff
@@ -1,50 +1,49 @@
+mod access;
+
 use std::fmt;
 use std::num::ParseIntError;
 
-use crate::{Reflect, ReflectMut, ReflectRef, VariantType};
+use crate::Reflect;
+use access::{Access, AccessRef};
 use thiserror::Error;
 
-/// An error returned from a failed path string query.
+type ParseResult<T> = Result<T, ReflectPathParseError>;
+
+/// An error specific to accessing a field/index on a `Reflect`.
 #[derive(Debug, PartialEq, Eq, Error)]
-pub enum ReflectPathError<'a> {
-    #[error("expected an identifier at index {index}")]
-    ExpectedIdent { index: usize },
-    #[error("the current struct doesn't have a field with the name `{field}`")]
-    InvalidField { index: usize, field: &'a str },
-    #[error("the current struct doesn't have a field at the given index")]
-    InvalidFieldIndex { index: usize, field_index: usize },
-    #[error("the current tuple struct doesn't have a field with the index {tuple_struct_index}")]
-    InvalidTupleStructIndex {
-        index: usize,
-        tuple_struct_index: usize,
-    },
-    #[error("the current tuple doesn't have a field with the index {tuple_index}")]
-    InvalidTupleIndex { index: usize, tuple_index: usize },
-    #[error("the current struct variant doesn't have a field with the name `{field}`")]
-    InvalidStructVariantField { index: usize, field: &'a str },
-    #[error("the current tuple variant doesn't have a field with the index {tuple_variant_index}")]
-    InvalidTupleVariantIndex {
-        index: usize,
-        tuple_variant_index: usize,
-    },
-    #[error("the current list doesn't have a value at the index {list_index}")]
-    InvalidListIndex { index: usize, list_index: usize },
+#[error(transparent)]
+pub struct AccessError<'a>(access::Error<'a>);
+
+/// A parse error for a path string query.
```

Comment:
Nit: Maybe remove `query` to make it clear that this isn't related to a failed query, but a failed parsing.
```suggestion
/// A parse error for a path string.
```